docs/technical-reference/Mutation Orchestration Design.md

```diff
@@ -0,0 +1,320 @@
+#### MutateAtStatementLevelOrchestrator
+`MutateAtStatementLevelOrchestrator` class deals with (expression level) syntax constructs which mutations should be controlled at statement level, because they must be controlled via an `if` statement.
+The constructor accepts a _predicate_ which is used for finer identification of SyntaxNodes.
+This class implements:
+- `CanHandle(...)`: forwards the request to the predicate (if any)
+- `StoreMutations(...)`: stores mutation at the next higher statement level.
+This class is used for:
+- `Post/PrefixUnaryExpressionSyntax` (\++/\—): when part of a for statement or an expression statement
+- `InitializerExpressionSyntax`: for non empty array declarations.
+- `AssignmentExpressionSyntax`( x=value) as they  cannot be part of a ternary operation.
+
+#### MemberDefinitionOrchestrator
+`MemberDefinitionOrchestrator<T>` is a generic (base) class to help implement orchestrator for SyntaxNode that define class members. It implements:
+- `PrepareContext(...)`: creates a `MutantControl.Member` context.
+- `RestoreContext()`: leaves the current context
+
+#### BaseFunctionOrchestrator
+`BaseFunctionOrchestrator<T>` is an abstract base class helping orchestrate methods, functions (locals and anonymous) as well as accessors. This class will convert  from expression body to block body form if any mutation requires it. This class also implements `IInstrumentCode` and provides the adequate roll back logic for the compilation phase. It inherits from `MemberDefinitionOrchestrator<T>`.
+Child classes must implement abstract methods that abstract some common operations:
+- `ParameterList(T)`: must return the `ParameterListSyntax` instance listing all the provided method’s parameters. Returns an empty list if there are none.
+- `ReturnType(T)`:  must return the `TypeSyntax` for the method’s returning type.  Should return `null` If the type cannot be determined (anonymous lambdas). One can use `RoslynHelper.VoidTypeSyntax()` if the method is void , explicitly or implicitly (constructors for example).
+- `GetBodies(T)`: must return the body for the method. Return value is a tuple with the block body as first item and the expression body as the second. At least one of them should be null
+- `SwitchToThisBodies(...)`: Must replace the method body with the provided body (either block or expression). This method is used to perform the expression to block body conversion and for roll back as well. Note that implementations assumes one of the provided bodies is null (as a function cannot have both an expression body and block statement body;
+Orchestrator’s overrides:
+- `InjectMutations(...)`: if the body is in expression form and there are mutations that have not been injected (because they require statement or block level control), converts the body to block form and inject those mutations. If the body is in block form, injects out parameters initialization to default and add an ending `return`  statement if it looks helpful.
```

Comment:
```suggestion
- `InjectMutations(...)`: if the body is in expression form and there are mutations that have not been injected (because they require statement or block level control), converts the body to block form and injects those mutations. If the body is in block form, injects out parameters initialization to default and adds an ending `return`  statement if it looks helpful.

```